File: 3/main.rb
# sliding window optimized (hash table)

# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
    if s.empty?
        return 0
    end
    
    map = {}
    result = 1
    left = -1
    s.size.times do |i|
        if map.has_key?(s[i]) && map[s[i]] > left
            left = map[s[i]]
        end
        
        result = [result, i - left].max
        map[s[i]] = i
    end
    
    result
end

File: 3/main.c
// sliding window optimized (ASCII 128)

#define MAGIC 128

inline int max(const int a, const int b) {
    return a > b ? a : b;
}

int lengthOfLongestSubstring(char * s){
    int n = strlen(s), result = 0;
    if(s == NULL) {
        return result;
    }
    int index[MAGIC] = { -1 };
    
    for(int j = 0, i = 0; j < n; j++) {
        i = max(i, index[s[j]]);
        result = max(result, j - i + 1);
        index[s[j]] = j + 1;
    }
    
    return result;
}

